src/emmy/expression/compile.cljc

```diff
@@ -124,232 +123,6 @@
-  nondeterministic order of hash maps inside this function won't guarantee a
-  consistent variable naming convention in the returned function. For tests, set
-  `:deterministic? true`."
-  ([expr continue] (extract-common-subexpressions expr continue {}))
-  ([expr continue {:keys [symbol-generator deterministic?]
-                   :or {symbol-generator sortable-gensym}}]
-   (let [sort (if deterministic?
-                (partial sort-by (comp str vec first))
-                identity)]
-     (loop [x         expr
-            expr->sym {}]
-       (let [expr->count (expr-frequencies x expr->sym)
-             new-syms    (into {} (for [[k v] (sort expr->count)
-                                        :when (> v 1)]
-                                    [k (symbol-generator)]))]
-         (if (empty? new-syms)
-           (let [sym->expr (-> (set/map-invert expr->sym)
-                               (discard-unreferenced-syms x))]
-             (continue x (sort-by key sym->expr)))
-           (let [expr->sym' (merge expr->sym new-syms)]
-             (recur (w/postwalk-replace expr->sym' x)
-                    expr->sym'))))))))
-
-;; This final wrapper function invokes `extract-common-subexpressions` to turn a
-;; symbolic expression a new, valid Clojure(script) form that uses a `let`
-;; binding to bind any common subexpressions exposed during the above search.
-;;
-;; If there are no common subexpressions, `cse-form` will round-trip its input.
-
-(defn cse-form
-  "Given a symbolic expression `expr`, returns a new expression potentially
-  wrapped in a `let` binding with one binding per extracted common
-  subexpression.
-
-  ## Optional Arguments
-
-  `:symbol-generator`: side-effecting function that returns a new, unique symbol
-  on each invocation. These generated symbols are used to create unique binding
-  names for extracted subexpressions. `sortable-gensym` by default.
-
-  NOTE that the symbols should appear in sorted order! Otherwise we can't
-  guarantee that the binding sequence won't contain entries that reference
-  previous entries, resulting in \"Unable to resolve symbol\" errors.
-
-  `:deterministic?`: if true, the function will order the let-binding contents
-  by sorting the string representations of each term before assignment. If false
-  the function won't guarantee a consistent variable naming convention in the
-  returned function. For tests, we recommend `:deterministic? true`."
-  ([expr] (cse-form expr {}))
-  ([expr opts]
-   (letfn [(callback [new-expression bindings]
-             (let [n-bindings (count bindings)]
-               (if (pos? n-bindings)
-                 (let [binding-vec (into [] cat bindings)]
-                   (log/info
-                    (format "common subexpression elimination: %d expressions" n-bindings))
-                   `(let ~binding-vec
-                      ~new-expression))
-                 new-expression)))]
-     (extract-common-subexpressions expr callback opts))))
```

Comment:
I wonder if there is some way (not now, down the road) to use this feature in the TeX renderer? Is there some standard way to describe local bindings in TeX?

Reply:
It could be done. In fact, it might even be cool. Something like (to borrow a Haskellism):
$$
\bf{let}
\qquad A_1 := x+y\cr
\qquad A_2 := \sin(x)\cr
\bf{in}
\qquad A_1^2 + A_2
$$
I don't know if that would actually render. I was sort of hoping GitHub would try to